pkg/resource/deploy/source_eval.go

```diff
@@ -1008,6 +1021,75 @@ func inheritFromParent(child resource.Goal, parent resource.Goal) *resource.Goal
 	return &goal
 }
 
+type sourcePositions struct {
+	projectRoot string
+}
+
+func newSourcePositions(projectRoot string) *sourcePositions {
+	if projectRoot == "" {
+		projectRoot = "/"
+	} else {
+		contract.Assertf(filepath.IsAbs(projectRoot), "projectRoot is not an absolute path")
+		projectRoot = filepath.Clean(projectRoot)
+	}
+	return &sourcePositions{projectRoot: projectRoot}
+}
+
+func (s *sourcePositions) parseSourcePosition(raw *pulumirpc.SourcePosition) (string, error) {
+	if raw == nil {
+		return "", nil
+	}
+
+	if raw.Line <= 0 {
+		return "", fmt.Errorf("invalid line number %v", raw.Line)
+	}
+
+	col := ""
+	if raw.Column != 0 {
+		if raw.Column < 0 {
+			return "", fmt.Errorf("invalid column number %v", raw.Column)
+		}
+		col = "," + strconv.FormatInt(int64(raw.Column), 10)
+	}
+
+	posURL, err := url.Parse(raw.Uri)
+	if err != nil {
+		return "", err
+	}
+	if posURL.Scheme != "file" {
+		return "", fmt.Errorf("unrecognized scheme %q", posURL.Scheme)
+	}
+
+	file := filepath.FromSlash(posURL.Path)
+	if !filepath.IsAbs(file) {
+		return "", fmt.Errorf("source positions must include absolute paths")
+	}
+	rel, err := filepath.Rel(s.projectRoot, file)
+	if err != nil {
+		return "", fmt.Errorf("making relative path: %w", err)
+	}
+
+	posURL.Scheme = "project"
+	posURL.Path = "/" + filepath.ToSlash(rel)
+	posURL.Fragment = fmt.Sprintf("%v%s", raw.Line, col)
+
+	return posURL.String(), nil
+}
+
+type hasSourcePosition interface {
```

Comment:
Maybe a comment that this interface is so we can pass "Register"/"Read"/etc objects into `getFromRequest` uniformly.

Reply:
Will do